File: packages/engine/src/lifecycle/ttl.ts
import { getContext } from "iii-sdk"
import type { StateKV } from "../state/kv.js"
import { SCOPES } from "../state/schema.js"
import { getDocker } from "../docker/client.js"
import { incrementExpired } from "../functions/metrics.js"
import type { Sandbox } from "../types.js"

export function registerTtlSweep(sdk: any, kv: StateKV) {
  sdk.registerFunction(
    { id: "lifecycle::ttl-sweep", description: "Kill expired sandboxes" },
    async (): Promise<{ swept: number }> => {
      const ctx = getContext()
      const sandboxes = await kv.list<Sandbox>(SCOPES.SANDBOXES)
      const now = Date.now()
      let swept = 0

      for (const sandbox of sandboxes) {
        if (sandbox.expiresAt <= now) {
          ctx.logger.info("Expiring sandbox", { id: sandbox.id })
          try {
            const container = getDocker().getContainer(`iii-sbx-${sandbox.id}`)
            await container.stop().catch(() => {})
            await container.remove({ force: true })
          } catch {}
          await kv.delete(SCOPES.SANDBOXES, sandbox.id)
          incrementExpired()
          swept++
        }
      }

      if (swept > 0) ctx.logger.info("TTL sweep complete", { swept })
      return { swept }
    },
  )

  sdk.registerFunction(
    { id: "lifecycle::health", description: "Health check" },
    async (): Promise<{ status: string; uptime: number }> => {
      return {
        status: "healthy",
        uptime: Math.floor(process.uptime()),
      }
    },
  )
}

File: packages/worker/src/lifecycle/ttl.rs
use bollard::container::RemoveContainerOptions;
use bollard::Docker;
use iii_sdk::III;
use serde_json::{json, Value};
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::state::{scopes, StateKV};
use crate::types::Sandbox;

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as u64
}

pub fn register(bridge: &Arc<III>, dk: &Arc<Docker>, kv: &StateKV) {
    {
        let kv = kv.clone();
        let dk = dk.clone();
        bridge.register_function("lifecycle::ttl-sweep", move |_input: Value| {
            let kv = kv.clone();
            let dk = dk.clone();
            async move {
                let sandboxes: Vec<Sandbox> = kv.list(scopes::SANDBOXES).await;
                let now = now_ms();
                let mut swept: u64 = 0;

                for sandbox in &sandboxes {
                    if sandbox.expires_at <= now {
                        let cn = format!("iii-sbx-{}", sandbox.id);
                        let _ = dk.stop_container(&cn, None).await;
                        let _ = dk
                            .remove_container(
                                &cn,
                                Some(RemoveContainerOptions {
                                    force: true,
                                    ..Default::default()
                                }),
                            )
                            .await;
                        let _ = kv.delete(scopes::SANDBOXES, &sandbox.id).await;
                        swept += 1;
                    }
                }

                Ok(json!({ "swept": swept }))
            }
        });
    }

    {
        bridge.register_function("lifecycle::health", move |_input: Value| {
            async move { Ok(json!({ "status": "healthy" })) }
        });
    }
}
